### PR TITLE
Implement network failover loadbalancing

### PR DESCRIPTION
**Pull request by hzxuzhonghu**
Implement https://github.com/istio/api/pull/1978

TODO: add some tests when it makes sense

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

Pull Request Attributes

Please check any characteristics that apply to this pull request. 

[ ] Does not have any [user-facing](https://github.com/istio/istio/tree/master/releasenotes#when-to-add-release-notes) changes. This may include CLI changes, API changes, behavior changes, performance improvements, etc.

### CODE REVIEW

**Comment by hzxuzhonghu** [comment]
@costinm @howardjohn @ramaraochavali @nrjpoddar

File: pilot/pkg/networking/core/v1alpha3/loadbalancer/loadbalancer.go [hunk 1]
```diff
@@ -56,23 +57,70 @@ func GetLocalityLbSetting(
...
 
-func ApplyLocalityLBSetting(
-	locality *core.Locality,
+func GetNetworkLbSetting(
+	mesh *v1alpha3.NetworkLoadBalancerSetting,
+	destrule *v1alpha3.NetworkLoadBalancerSetting,
+) *v1alpha3.NetworkLoadBalancerSetting {
+	var enabled bool
+	// Locality lb is enabled if its not explicitly disabled in mesh global config
+	if mesh != nil && (mesh.Enabled == nil || mesh.Enabled.Value) {
```
Comment (ramaraochavali) [inline pilot/pkg/networking/core/v1alpha3/loadbalancer/loadbalancer.go:66]: nit: I think you can assign enabled = mesh.Enabled.Value same for Drs below. Code might be simplified

File: pilot/pkg/networking/core/v1alpha3/loadbalancer/loadbalancer.go [hunk 2]
```diff
@@ -56,23 +57,70 @@ func GetLocalityLbSetting(
...
 	// one of Distribute or Failover settings can be applied.
 	if localityLB.GetDistribute() != nil {
-		applyLocalityWeight(locality, loadAssignment, localityLB.GetDistribute())
+		applyLocalityWeight(proxyLocality, loadAssignment, localityLB.GetDistribute())
 		// Failover needs outlier detection, otherwise Envoy will never drop down to a lower priority.
 		// Do not apply default failover when locality LB is disabled.
-	} else if enableFailover && (localityLB.Enabled == nil || localityLB.Enabled.Value) {
-		applyLocalityFailover(locality, loadAssignment, localityLB.GetFailover())
+	} else if enableFailover {
+		if networkLB == nil {
```
Comment (ramaraochavali) [inline pilot/pkg/networking/core/v1alpha3/loadbalancer/loadbalancer.go:115]: This condition does not match with the call.

Comment (ramaraochavali) [inline pilot/pkg/networking/core/v1alpha3/loadbalancer/loadbalancer.go:115]: Is it guaranteed that if networkLB is nil, locality will have values?

File: pilot/pkg/networking/core/v1alpha3/loadbalancer/loadbalancer.go [hunk 3]
```diff
@@ -166,6 +214,96 @@ func applyLocalityFailover(
...
+	loadAssignment *endpoint.ClusterLoadAssignment,
+	proxyNetwork string,
+	networkFailover []*v1alpha3.NetworkLoadBalancerSetting_Failover) {
+	// key is priority, value is the index of the LocalityLbEndpoints in ClusterLoadAssignment
+	priorityMap := map[int][]int{}
+
+	// 1. calculate the LocalityLbEndpoints.Priority compared with proxy network
+	for i, localityEndpoint := range loadAssignment.Endpoints {
+		// all the LbEndpoints belong to same network as we have filtered in ep_filter.go
+		epNetwork := util.IstioMetadata(localityEndpoint.LbEndpoints[0], "network")
```
Comment (ramaraochavali) [inline pilot/pkg/networking/core/v1alpha3/loadbalancer/loadbalancer.go:231]: nit: define network metadata name as const some where

**Comment by hzxuzhonghu** [comment]
close in favor of #33047

**Pull request closed without merging**